[Feat]: Added a debug overlay

Vanshikaxxa opened the pull request:
This PR introduces an FPS counter and a stats rendering system to display real-time FPS, entity count, and resolution. The system utilizes a pixel-based font rendering method.

### Changes
- **Added FPS calculation** (`calculate_fps`) using `Unix.gettimeofday()`
- **Implemented character rendering** (`draw_char`) for FPS display using a pixel grid  
- **Created string rendering function** (`draw_string`) to display text on the screen  
- **Added `render_stats` function** to render FPS, entity count, and resolution  
- **Declared necessary functions in `overlay.mli`** for module interface consistency  

### Implementation Details  
- **FPS calculation:** Updates once per second by tracking frame count.  
- **Font rendering:** Uses a pixel-mapped approach for letters and numbers.  
- **Toggle key:** Defined `stats_toggle_key` (`Key.F`) to enable/disable stats display.

mdales reviewed (changes requested):
This is a great start, thank you! I love that you hard coded your own font into this :D I'm sorry we haven't already done #48. I also like that you've done it all within Claudius's own graphics routines.

Some overall thoughts though:

* This needs to be integrated into the main runloop in base.ml, so it's always there for every application.
* I appreciate what you're trying to do with the entity_count, but not all apps will have an entity count currently, so I'd stick to things that you can infer about any program (which is important when this is integrated with the run loop in base.ml

So it's a great start, but it needs to be hooked up so that all programs get to use this feature without having to call it themselves! Does that make sense?

Vanshikaxxa commented:
Hey @mdales,
I’ve moved the overlay to the base module and removed the entity count function. However, the overlay is now lagging when it's rendered. Do you have any suggestions for optimizing it?

mdales commented:
I had a play with this, this is going to be so useful! I can already see that I get a very different framerate for Claudius depending on my monitor and power settings!

On my external display, or when my laptop is on power save mode I get 30 FPS:

<img width="752" alt="Screenshot 2025-04-10 at 09 04 52" src="https://github.com/user-attachments/assets/8b27f491-24a6-467c-ab0b-18a2916c5c95" />

And on my internal laptop display when it is plugged in I get 60 FPS:

<img width="752" alt="Screenshot 2025-04-10 at 09 04 46" src="https://github.com/user-attachments/assets/a21c2747-5db4-4ad1-834f-3f841923f3dc" />

Very cool!

Now, there is a bug in your current implementation I had to quickly fix: I see you've put in a bunch of caching to try limit the impact, I assume because you're having issues with performance as you mentioned, but you've accidentally put the `calculate_fps` function into the "don't refresh unless half a second has passed" section, and so you only get 2 FPS with your branch :)

In terms of performance impact: I guess the first question is are you using WSL, in which case have you hit #55 - it seems that SDL, the graphics library that Claudius is based on, does have performance issues on WSL, and there's some notes been added to the README about that.

Can you talk a bit more about what you're seeing? Unfortunately for me your branch seems to be working well, which makes it hard for me to assist - it's always annoying when there a bugs or problems that you can't reproduce! If you use the day1 example from the examples, what is the FPS you get with the overlay shown, and what is the FPS without it (you can just use Printf to log it - SDL stops prints whilst running, but they should all magically appear when you quit). 

I guess I'm just working out: is this just Claudius being slow generally, or is it definitely to do with your change? At the moment your change is working okay, and whilst I've not read the code in detail, I didn't spot you doing anything obviously going to cause issues in there.

If I get chance later today I'll fire up my Windows machine and test it there (I did these on my M1 MacBook Pro, which whilst now a little old, is still quite a fast computer, so might be a bad machine for testing performance issues on).

But in general, I think you should tidy this up so I can do a proper review (as in, fix the merge conflicts), as this is a great addition. The one thing I did spot, is that you updated `run_functional`, but that j just calls the regular `run` inside it, so your code will be invoked eventually, you don't need to update that.
